io/pci/dlpar.py

```diff
@@ -367,6 +381,101 @@ def changehwres_sriov(self, server, operation, lpar_id, adapter_id,
+
+        def fc_recovery_check():
+            """
+            Checks if the block device adapter is recovers all its disks/paths
+            properly after hotplug of adapter.
+            Returns True if all disks/paths back online after adapter added
+            Back, else False.
+            """
+            def is_path_online():
+                path_stat = list(multipath.get_path_status(curr_path))
+                if path_stat[0] != 'active' or path_stat[2] != 'ready':
+                    return False
+                return True
+
+            curr_path = ''
+            err_disks = []
+            disks = pci.get_disks_in_pci_address(self.pci_device)
+            for disk in disks:
+                curr_path = disk.split("/")[-1]
+                self.log.info("curr_path=%s" % curr_path)
+                if not wait.wait_for(is_path_online, timeout=10):
+                    self.log.info("%s failed to recover after add" % disk)
+                    err_disks.append(disk)
+
+            if err_disks:
+                self.log.info("few paths failed to recover : %s" % err_disks)
+                return False
+            return True
+
+        def net_recovery_check():
+            """
+            Checks if the network adapter fuctionality like ping/link_state,
+            after adapter added back.
+            Returns True on propper Recovery, False if not.
+            """
+            self.log.info("entering the net recovery check")
+            local = LocalHost()
+            iface = pci.get_interfaces_in_pci_address(self.pci_device, 'net')
+            networkinterface = NetworkInterface(iface[0], local)
+            if wait.wait_for(networkinterface.is_link_up, timeout=120):
+                if networkinterface.ping_check(self.peer_ip, count=5) is None:
+                    self.log.info("inteface is up and pinging")
+                    return True
+            return False
+
+        def nvme_recovery_check():
+            '''
+            Checks if the nvme adapter functionality like all namespaces are
+            up and running or not after adapter is recovered
+            '''
+            err_ns = []
+            current_namespaces = nvme.get_current_ns_ids(self.contr_name)
+            if current_namespaces == self.ns_list:
+                for ns_id in current_namespaces:
+                    status = nvme.get_ns_status(self.contr_name, ns_id)
+                    if not status[0] == 'live' and status[1] == 'optimized':
+                        err_ns.append(ns_id)
+            else:
+                self.log.info("following ns not back listing after hot_plug" %
+                              (self.ns_list - current_namespaces))
+                return False
+
+            if err_ns:
+                self.log.info(f"following namespaces not recovered ={err_ns}")
+                return False
+            return True
+
+        if wait.wait_for(is_added, timeout=30):
+            time.sleep(45)
```

Comment:
one last comment if wait is given than why sleep ?

Reply:
this is for splitter limitation. so added extra sleep.
we are expecting enhancement for splitter. so we can remove it later.